packages/commons/src/auth/authorization.ts

```diff
@@ -0,0 +1,210 @@
+      ? MaintenanceAuthData
+      : never)
+  // If there are user roles in the array, add UIAuthData
+  | ([ExtractUserRoles<AdmittedRoles>] extends [never] ? never : UIAuthData);
+
+export function validateAuthorization<
+  AdmittedRoles extends NonEmptyArray<AuthRole>
+>(
+  ctx: AppContext,
+  admittedAuthRoles: AdmittedRoles
+): asserts ctx is AppContext<AllowedAuthData<AdmittedRoles>> {
+  const { authData } = ctx;
+
+  match(authData)
+    .with(
+      { systemRole: P.union("m2m", "internal", "maintenance") },
+      ({ systemRole }) => {
+        const admittedSystemRoles: SystemRole[] =
+          admittedAuthRoles.filter(isSystemRole);
+        if (!admittedSystemRoles.includes(systemRole)) {
+          throw unauthorizedError(
+            `Invalid role '${systemRole}' for this operation`
+          );
+        }
+      }
+    )
+    .with({ systemRole: undefined }, (authData) => {
+      const admittedUserRoles: UserRole[] =
+        admittedAuthRoles.filter(isUserRole);
+
+      if (admittedUserRoles.length === 0) {
+        throw unauthorizedError(
+          `Invalid empty user roles specified for this operation`
+        );
+      }
+
+      if (!hasAtLeastOneUserRole(authData, admittedUserRoles)) {
+        throw unauthorizedError(
+          `Invalid roles ${JSON.stringify(
+            authData.userRoles
+          )} for this operation`
+        );
+      }
+    })
+    .exhaustive();
+}
+
+export function hasAtLeastOneUserRole(
+  authData: AuthData,
+  admittedUserRoles: ReadonlyArray<UserRole>
+): boolean {
+  return (
+    isUiAuthData(authData) &&
+    authData.userRoles.some((role: UserRole) =>
+      admittedUserRoles.includes(role)
+    )
+  );
+}
+
+function isUiAuthData(authData: AuthData): authData is UIAuthData {
+  return match(authData)
+    .with({ systemRole: undefined }, () => true)
+    .with(
+      { systemRole: P.union("m2m", "internal", "maintenance") },
+      () => false
+    )
+    .exhaustive();
+}
+function isSystemRole(role: AuthRole): role is SystemRole {
+  return match(role)
+    .with("m2m", "internal", "maintenance", () => true)
+    .with("admin", "security", "api", "support", () => false)
+    .exhaustive();
+}
+
+function isUserRole(role: AuthRole): role is UserRole {
+  return match(role)
+    .with("admin", "security", "api", "support", () => true)
+    .with("m2m", "internal", "maintenance", () => false)
+    .exhaustive();
+}
+
+// EXAMPLE USAGES -- UNCOMMENT EACH ENTIRE EXAMPLE TO TEST IT:
+
+// EXAMPLE 1 - UI with one user role
+// const mockContext = {} as AppContext;
+// /* eslint-disable @typescript-eslint/no-unused-vars */
+// validateAuthorization(mockContext, ["admin"]); // compiles
+// const systemRole = mockContext.authData.systemRole; // compiles and is "ui"
+// const orgId = mockContext.authData.organizationId; // compiles
+// const userId = mockContext.authData.userId; // compiles
+// -------------------------------------------------------
```

Comment:
readability can be increased using comment block

```suggestion
/* EXAMPLE 1 - UI with one user role
   const mockContext = {} as AppContext;
   // eslint-disable @typescript-eslint/no-unused-vars */
   validateAuthorization(mockContext, ["admin"]); // compiles
   const systemRole = mockContext.authData.systemRole; // compiles and is "ui"
   const orgId = mockContext.authData.organizationId; // compiles
   const userId = mockContext.authData.userId; // compiles
 --------------------------------------------------------- */
```

Reply:
It becomes more readable but less usable - it's harder to uncomment with the shortcut, and also there is `/* eslint-disable @typescript-eslint/no-unused-vars */` interfering with the block comment. In this way I can just select the entire example, uncomment using `CMD + /` and see it working